controller/src/main/java/io/pravega/controller/fault/ControllerClusterListener.java

```diff
@@ -42,32 +44,48 @@
     private final ExecutorService executor;
     private final Optional<ControllerEventProcessors> eventProcessorsOpt;
     private final TaskSweeper taskSweeper;
+    private final Optional<TxnSweeper> txnSweeperOpt;
 
     public ControllerClusterListener(final Host host, final Cluster cluster,
                                      final Optional<ControllerEventProcessors> eventProcessorsOpt,
                                      final TaskSweeper taskSweeper,
+                                     final Optional<TxnSweeper> txnSweeperOpt,
                                      final ExecutorService executor) {
         Preconditions.checkNotNull(host, "host");
         Preconditions.checkNotNull(cluster, "cluster");
         Preconditions.checkNotNull(executor, "executor");
         Preconditions.checkNotNull(eventProcessorsOpt, "eventProcessorsOpt");
         Preconditions.checkNotNull(taskSweeper, "taskSweeper");
+        Preconditions.checkNotNull(txnSweeperOpt, "txnSweeperOpt");
 
         this.objectId = "ControllerClusterListener";
         this.host = host;
         this.cluster = cluster;
         this.executor = executor;
         this.eventProcessorsOpt = eventProcessorsOpt;
         this.taskSweeper = taskSweeper;
+        this.txnSweeperOpt = txnSweeperOpt;
     }
 
     @Override
-    protected void startUp() throws Exception {
+    protected void startUp() throws InterruptedException, Exception {
         long traceId = LoggerHelpers.traceEnter(log, objectId, "startUp");
         try {
             log.info("Registering host {} with controller cluster", host);
             cluster.registerHost(host);
 
+            if (eventProcessorsOpt.isPresent()) {
+                // Await initialization of eventProcesorsOpt
```

Comment:
1. Yes, cluster listener will wait on event processors to come up.
2. No, Cluster listener is not used for populating host store entries. Host store is the first thing started by ControllerServiceStarter. There is no cyclic dependency here.